proposals/text/5540-llm-support-2.0.md

```diff
@@ -0,0 +1,309 @@
+3. `Cohere`
+4. `HuggingFaceInferenceLLM`
+5. `HuggingFaceLocalLLM`
+6. `GPT3`
+6. `AzureGPT3`?
+7. `Bard` ?
+8. `SagemakerLLM`
+
+Plus one more for any other inference hosting/library that may appear in the future.
+
+## Prompt Builder
+
+In Haystack 1.x, prompts fetching and rendering is carried out by `PromptTemplate`. In 2.0, we rather make a separate `Prompts` component to handle this process.
+
+Due to the dynamic nature of prompt templates, the `Prompts.run()`  method takes `kwargs`, which contains all the variables that will be filled in the template. However, for this component to work with Canals, we need to know in advance which values this dict will contain: therefore, we need the users to specify in the `__init__` of the component which parameters to expect in the template.
+
+Keep in mind that such parameters names **cannot be changed at runtime**.
+
+We foresee `Prompts` to output a list of prompts, not necessarily just one, as currently  some of our existing prompt templates produce several output prompts when rendered.
+
+Draft I/O for `Prompts`:
+
+```python
+@component
+class Prompts:
+
+    def __init__(self, template_variables: Union[str, Path]):
+        self.template_variables = template_variables
+		component.set_input_parameters(**{var: Any for var in template_variables})
+
+  	@component.output_types(prompts=List[str])
+    def run(self, template: str, **kwargs):
+        self.template_text = # Load the template
+
+        variables = # extracts the variables from the template text
+        if variables != self.template_variables:
+            raise ValueError()
+
+        # Render the template using the variables
+        return {"prompts": prompts}
+```
+
+### **Why we need to specify the variables in `__init__`?**
+
+The design above derives from one Canals limitation: component’s sockets need to be all known the latest at `__init__` time, in order for the connections to be made and validated. Therefore, we need to know all the prompt variables before building the pipelines, because the prompt variables are inputs of the `run()` method.
+
+However, earlier iterations of Canals did support so-called “true variadic” components: components that do not need to know what they will be connected to, and build the input sockets at need. Such components of course lack input validation, but enable usecases like the above.
+
+If we decide that Canals should support again such components, we would be able to rewrite `Prompts` to take a prompt as its input parameter and just accept any other incoming input, on the assumption that users knows that they’re doing.
+
+For example:
+
+```python
+@component
+class Prompts:
+
+	@variadic_input
+  	@component.output_types(prompts=List[str])
+    def run(self, template: Union[str, Path], **kwargs):
+	    # ... loads the template ...
+        # ... render the prompts ...
+        return {"prompts": prompts}
+```
+
+### Why a separate `Prompts` component at all?
```

Comment:
Will `Prompts` be tokenizer aware?

We might need to count input tokens from the prompt or we might want to truncate some variables in the prompt but not others (e.g. truncate documents but not the instructions), will that we possible?

Reply:
> We might need to count input tokens from the prompt or we might want to truncate some variables in the prompt but not others (e.g. truncate documents but not the instructions), will that we possible?

Alright that's very interesting :eyes: My answer would be no, `Prompts` is not tokenizer aware. However, your usecase is very valid, and the current design wouldn't support it right away.

I'll rewrite pros and cons of this situation in a separate document, take a decision again, and update the proposal. Thank you catching it early! :pray:

Reply:
Thanks for taking this into consideration! I also just wanted to chime in and say that the situation Mathis described is something that Sol has run into a lot and would be extremely useful (almost needed) to make RAG QA more reliable in production.

Reply:
Why should the prompt component be aware of the tokenizer? Couldn't it be better to have a separate component that does just that? Or maybe it's something for the prompt renderer component. 🤔 

I envision the prompt components to be kinda simple, they get the prompt as input and send it to whoever model they talk too. Adding a tokenizer in the mix seems overkill to me.

Reply:
> Or maybe it's something for the prompt renderer component.

I see, I don't think I fully understood the flow of the prompt from my initial read of the proposal. From what I understand in the proposal the rendered component would be where the final prompt is constructed given the variables from the invocation context. Is that correct?

If so, I think having this component be tokenizer aware would satisfy the use case Mathis and I are thinking of. Although from the current version of the proposal it looks like the rendering will still occur in the Prompt class? If that's the case will we pass a rendering component to the Prompt component?

Reply:
I'm confused now, there will be a separate prompt renderer?

What I don't like about the prompt-anything to be tokenizer aware is that the user actually needs to know what the tokenizer is that should be used with a specific model. In the current design they don't need to worry about the tokenizer at all.

Reply:
As a side note:

The more we're talking about another small component that should be used instead, the more interested I am in that component bundling layer that was mentioned a few times. I get it from a design standpoint that each component should do one small thing but we will need to compose these components into simpler abstractions for the end user. Nobody wants to create pipelines with dozens of components. Is there any plan on an upcoming proposal for this layer yet?

Reply:
> I'm confused now, there will be a separate prompt renderer?

That would be what's discussed in [this section](https://github.com/deepset-ai/haystack/pull/5540/files#diff-47e76fa36873ebd950a2ca709e6da056a4a42e6f80d1cda59a8fdb68886ac58cR150), the name of the class is not descriptive enough I think. We should change it to make the proposal less confusing.

> The more we're talking about another small component that should be used instead, the more interested I am in that component bundling layer that was mentioned a few times. I get it from a design standpoint that each component should do one small thing but we will need to compose these components into simpler abstractions for the end user. Nobody wants to create pipelines with dozens of components. Is there any plan on an upcoming proposal for this layer yet?

Bundling components can always be done at a later time if we see it's more beneficial. As of now we see more benefits in keeping them smaller. We separate concerns, make it easier and faster to create new components, less error prone logic, easier to tests, etc. 

I was doubtful about making the rendering of a prompt string template a separate component but in the end after some thinking I think it's the best decision for now.

Again ideally wrapping multiple components into a single one will be simple enough that any user could do it if they see the need to do so. As of now we prefer to create simple components at the cost of making Pipelines have more components.

Reply:
> Again ideally wrapping multiple components into a single one will be simple enough that any user could do it if they see the need to do so. As of now we prefer to create simple components at the cost of making Pipelines have more components.

I think @masci had said in other places that some of these wrappers would be provided by Haystack. If we leave it to the user completely, I'd be very concerned about the usability of v2.

Reply:
That depends on a case by case basis and it can be done later on. Wrappers are easy to build. 
I believe this is not part of the scope of this proposal.

Reply:
I asked because we are adding yet more components here. If it's simple, great! I'd just like to see end to end examples of common use cases with wrappers at some point. I think it's better to have these earlier than later.

But I agree, no need to discuss further here.

Reply:
Let's talk about this topic in here: https://www.notion.so/deepsetai/Separate-Prompt-rendering-component-or-not-3fa781f429d34685938428794aeeb786

Reply:
I'll resolve this one **only for readability purposes**. The discussion goes on on Notion!